app/server/lib/ActiveDoc.ts

```diff
@@ -987,8 +989,16 @@ export class ActiveDoc extends EventEmitter {
           throw new ApiError("Document is shutting down, archiving aborted", 500);
         }
         const file = await attachmentFileManager.getFile(attachment.fileIdent);
-        const fileHash = attachment.fileIdent.split(".")[0];
-        const name = `${fileHash}_${attachment.fileName}`;
+        const fileIdentParts = attachment.fileIdent.split(".");
+        const fileHash = fileIdentParts[0];
+        const fileIdentExt = path.extname(attachment.fileIdent);
+        const fileNameExt = path.extname(attachment.fileName);
+        const fileNameNoExt = path.basename(attachment.fileName, fileNameExt);
+        // We need to make sure the downloaded attachment's extension matches Grist's internal
+        // file extension, otherwise we can't recreate the file identifier when uploading.
+        // They might not match, as the upload process considers things like mime type when
+        // adding the extension to the file identifier.
+        const name = `${fileHash}_${fileNameNoExt}${fileIdentExt}`;
```

Comment:
Small tweak to the names of files in downloaded archives (see comment). Shouldn't make any difference 99% of the time, but covers an edge case.